### ISSUE TITLE
[Bug][compiler v2] Bytecode verifier error on mut ref equality check

### ISSUE THREAD

**Issue by vineethk** [aptos-labs/aptos-core#12151]
# 🐛 Bug

The following code successfully compiles with v1, but has a bytecode verifier error in v2 (when run with code in this branch, in a transactional test):

```
module 0xc0ffee::m {
    fun test(): bool {
        let x = 1;
        let r1 = &mut x;
        let r2 = &mut x;
        r1 == r2
    }
}
```

This issue is applicable after landing https://github.com/aptos-labs/aptos-core/pull/11892, with the original discussion here: https://github.com/aptos-labs/aptos-core/pull/11892#issuecomment-1957207608

**Comment by vineethk** [comment]
@wrwg Feel free to un-assign if not appropriate.

**Comment by wrwg** [comment]
Compiler v1 generates the following code:

```

- test(): bool /* def_idx: 0 */ {
- L0:	loc0: &mut u64
- L1:	loc1: &mut u64
- L2:	loc2: u64
- B0:
- 	0: LdU64(1)
- 	1: StLoc[2](loc2: u64)
- 	2: MutBorrowLoc[2](loc2: u64)
- 	3: StLoc[0](loc0: &mut u64)
- 	4: MutBorrowLoc[2](loc2: u64)
- 	5: StLoc[1](loc1: &mut u64)
- 	6: MoveLoc[0](loc0: &mut u64)
- 	7: FreezeRef
- 	8: MoveLoc[1](loc1: &mut u64)
- 	9: FreezeRef
- 	10: Eq
- 	11: Ret
```

v2 currently does not generate a `FreezeRef`, which seems to be the issue here. We likely should introduced the FreezeRef for parameters of type `&T` where the arg has type `&mut T` in the new ability_processor where we also generate `Copy` for arguments if needed.

The full error:

```
+ Error: compilation errors:
+  bug: BYTECODE VERIFICATION FAILED
+   ┌─ TEMPFILE:2:1
+   │  
+ 2 │ ╭ module 0x42::m {
+ 3 │ │     fun test(): bool {
+ 4 │ │         let x = 1;
+ 5 │ │         let r1 = &mut x;
+   · │
+ 8 │ │     }
+ 9 │ │ }
+   │ ╰─^ ICE failed bytecode verifier: VMError {
+     major_status: READREF_EXISTS_MUTABLE_BORROW_ERROR,
+     sub_status: None,
+     message: None,
+     exec_state: None,
+     location: Module(
+         ModuleId {
+             address: 0000000000000000000000000000000000000000000000000000000000000042,
+             name: Identifier(
+                 "m",
+             ),
+         },
+     ),
+     indices: [
+         (
+             FunctionDefinition,
+             0,
+         ),
+     ],
+     offsets: [
+         (
+             FunctionDefinitionIndex(0),
+             8,
+         ),
+     ],
+ }
```

**Comment by wrwg** [comment]
There is another bug which was surface here: `r1 == r2` should actually result in a borrow error if there is no freeze for the operands. This has been fixed in the latest commit of #12070 .
